src/main/java/ch/uzh/ifi/hase/soprafs21/security/JwtTokenAuthenticationFilter.java

```diff
@@ -0,0 +1,69 @@
+package ch.uzh.ifi.hase.soprafs21.security;
+
+import ch.uzh.ifi.hase.soprafs21.security.config.SecurityConstants;
+import ch.uzh.ifi.hase.soprafs21.service.JwtTokenUtil;
+import ch.uzh.ifi.hase.soprafs21.service.UserService;
+import io.jsonwebtoken.Claims;
+import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
+import org.springframework.security.core.context.SecurityContextHolder;
+import org.springframework.security.web.authentication.WebAuthenticationDetailsSource;
+import org.springframework.web.filter.OncePerRequestFilter;
+
+import javax.servlet.Filter;
+import javax.servlet.FilterChain;
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+
+public class JwtTokenAuthenticationFilter extends OncePerRequestFilter {
+
+    private JwtTokenUtil jwtTokenUtil;
+    private UserService userService;
+
+    public JwtTokenAuthenticationFilter(
+            JwtTokenUtil jwtTokenUtil,
+            UserService userService) {
+
+        this.jwtTokenUtil = jwtTokenUtil;
+        this.userService = userService;
+    }
+
+    @Override
+    protected void doFilterInternal(HttpServletRequest request, HttpServletResponse response, FilterChain filterChain) throws ServletException, IOException {
+        String header = request.getHeader(SecurityConstants.HEADER_STRING);
+
+        if(header == null || !header.startsWith(SecurityConstants.TOKEN_PREFIX)) {
+            response.setStatus(HttpServletResponse.SC_UNAUTHORIZED);
+            return;
+        }
+
+        String token = header.replace(SecurityConstants.TOKEN_PREFIX, "");
+
+        if(jwtTokenUtil.validateToken(token, SecurityConstants.SECRET)) {
+            Claims claims = jwtTokenUtil.getClaimsFromJWT(token, SecurityConstants.SECRET);
+            String emailId = claims.getSubject();
+
+            UsernamePasswordAuthenticationToken auth =
+                    userService.getUserByEmail(emailId)
+                            .map(userDetails -> {
+                                UsernamePasswordAuthenticationToken authentication =
```

Comment:
So if I understand well, we don't use anymore Google for authentication, except for the very first login?

Reply:
Yes, we are saving user and refresh token in DB during login and the client will send accessToken in the authorization header (Bearer token) for every other REST calls.

Reply:
I do not understand how the authentication will happened in case of for example: when the user uses different machine and there won't be this cookie anymore?